include/deal.II/base/function_lib.h

```diff
@@ -965,6 +1016,92 @@ namespace Functions
      * in all components.
      */
     const unsigned int selected;
+
+    /**
+     * Flag that controls wether we rescale the value when the radius changes.
+     */
+    bool integrate_to_one;
+
+    /**
+     * The reference integral value. Derived classes should specify what their
+     * integral is when @p radius = 1.0.
+     */
+    const double unitary_integral_value;
+
+    /**
+     * Current rescaling to apply the the cut-off function.
+     */
+    double rescaling;
+  };
+
+
+  /**
+   * Tensor product of CutOffFunctionBase objects.
+   *
+   * Instead of using the distance to compute the cut-off function, this class
+   * performs a tensor product of the same CutOffFunctionBase object in each
+   * coordinate direction.
+   *
+   * @ingroup functions
+   * @author Luca Heltai, 2019.
+   */
+  template <int dim>
+  class CutOffFunctionTensorProduct : public CutOffFunctionBase<dim>
+  {
+  public:
+    /**
+     * Construct an empty CutOffFunctionTensorProduct object.
+     *
+     * Before you can use this class, you have to call the set_base() method
+     * with a class derived from the CutOffFunctionBase object.
+     *
+     * If you try to use this class before you call the set_base() method,
+     * and exception will be triggered.
+     */
+    CutOffFunctionTensorProduct(
+      double             radius       = 1.0,
+      const Point<dim> & center       = Point<dim>(),
+      const unsigned int n_components = 1,
+      const unsigned int select       = CutOffFunctionBase<dim>::no_component,
+      const bool         integrate_to_one = false);
+
+    /**
+     * Initialize the class with an objet of type
+     * @tparam CutOffFunctionBaseType<1>.
+     */
+    template <template <int> class CutOffFunctionBaseType>
+    void
+    set_base();
```

Comment:
Wouldn't it be easier to pass in an object derived from `CutOffFunctionBase<dim>`, instead of creating one internally? This would seem like a more natural thing to do in my mind.

Reply:
I thought about this for some time, and my initial implementation had the possibility to pass a vector of objects and stored internally smart pointers to these objects. 

This has one drawback, that I'd like to avoid:
~~~
MyCutOffFunction<1> base;

set_base({&base, &base, &base});
~~~
this seemed apparently like  the correct thing to do, but would have given wrong results. In order for this to work, you'd have to create *dim* different objects externally in order for the function to do its job properly: when you set the center of the tensor product function, each object of the base sets its own center equal to the ith- coordinate of the center, and if you only construct one object externally, this won't work, since they'd all set the center to the last coordinate of the center.

I had in place exceptions that checked that the three objects were actually different, but then creating dim of these objects at run time was a little cumbersome, hence my choice here.

If we added a clone method, then this would work by simply cloning the base object internally dim times. Would you prefer this solution?

Reply:
I don't feel strongly. The downside of your current approach is that one cannot pass constructor arguments to the object so created. This could be avoided by making the function like `make_unique` that takes a variadic list of arguments that are passed to the function and that are passed on to the constructor of the object to be created.

You probably want to use `static_assert(std::is_base_of<...,....>)` to ensure that the type passed in is in fact derived from `CutOffFunctionBase`. (At the top of the implementation of the function.)

Reply:
> The downside of your current approach is that one cannot pass constructor arguments to the object so created. 

This is not really a downside. There are no free arguments you can set in the base, because these must match the ones used in the constructor of this class (n_components, selected component, and integrate_to_one). The center and the radius are given by the outer object (not by the single base), so these would not have any effect on the final object.

Reply:
ok